libcxx/include/valarray

```diff
@@ -819,6 +819,19 @@ public:
   _LIBCPP_HIDE_FROM_ABI valarray& operator<<=(const value_type& __x);
   _LIBCPP_HIDE_FROM_ABI valarray& operator>>=(const value_type& __x);
 
+  // The valarray overloads are needed for the types with an implicit
+  // conversion to valarray. For example slice_array.
+  _LIBCPP_HIDE_FROM_ABI valarray& operator*=(const valarray& __v) { return operator*= <valarray>(__v); }
```

Comment:
@ldionne There are a lot more functions to convert and a lot more tests to write. I like to discuss this approach before doing that work. It's the intention to do everything in one patch.

Reply:
Instead, I think we should make things like `gslice_array` & friends be `__val_expr`s and give them the necessary indexing operators (via a private name). That would allow for a much more efficient implementation of e.g. `valarray::operator*=` than if we need to create a temporary `valarray` first.